### PR TITLE
Add HSTU LayerNorm-Multiply-SiLU-Dropout CuTe DSL operation

### PR DESCRIPTION
**Pull request by z52527**
## Problem

HSTU training uses a fused LayerNorm–Multiply–SiLU–Dropout (LMSD) operation in both forward and backward. cuDNN Frontend currently has no dedicated implementation, so users must rely on an external Triton implementation and manage the saved tensors, compiled kernels, and backward workspaces separately.

At the production shape (`N=2,739,421`, `D=512`), the original Triton implementation measured 4.149 ms for forward and 2.912 ms for backward.

## What this PR does

Adds an experimental HSTU LMSD operation backed by CuTe DSL for BF16 tensors on SM10x GPUs.

- **Fused forward** — computes LayerNorm, affine transform, SiLU, multiplication, dropout, and packed mask generation. It returns `y`, `mean`, `rstd`, and an int8 mask containing three dropout decisions per element.
- **Fused backward** — consumes the saved tensors and computes `dx`, `du`, `dweight`, and `dbias`. It uses one persistent runtime-N kernel followed by a CuTe DSL dWeight/dBias reduction.
- **Explicit and convenience APIs** — provides precompiled, allocation-free `compile()` / `execute()` APIs together with cached Python convenience wrappers.
- **Large-tensor support** — uses 64-bit row rebasing for production strided tensors whose addressable storage exceeds 4 GiB.
- **Tests and benchmarks** — adds usage, fuzzer, integration, forward/backward microbenchmark, and forward-to-backward benchmark coverage.

The API is additive and currently supports BF16, `D=512`, and SM10x GPUs.

## Performance

Measured on SM107a with BF16 inputs, `N=2,739,421`, `D=512`, `u` row stride 2048, and memory clock locked at 4752 MHz:

| Direction | Original Triton | This PR | Speedup |
| --- | ---: | ---: | ---: |
| Forward | 4.149 ms | 2.103 ms | 1.97x |
| Backward | 2.912 ms | 1.866 ms | 1.56x |

The backward measurement includes the final dWeight/dBias reduction.

The historical Triton backward path used `compute_y=True`. The new backward API intentionally returns only gradients and does not recompute the unused forward output, so its improvement includes both the CuTe DSL implementation and removal of those unnecessary stores.

### Supported configurations

- **Sequence length (`N`)**: `1 <= N <= 4,194,304`
- **Feature dimension (`D`)**: `512` only
- **Data type**: BF16
- **GPU architecture**: SM10x
- **Tensor shapes**:
  - `x`, `u`: `[N, 512]`
  - `weight`, `bias`: `[512]`
  - `y`, `dY`: `[N, 1536]`

### Not supported

- `N = 0`
- `N > 4,194,304`
- `D != 512`
- Other data types
- Other GPU architectures

## Validation

Verified on SM107:

- Explicit forward and backward usage tests pass.
- `dx`, `du`, `dweight`, and `dbias` pass comparison against the FP32 PyTorch reference.
- The production-size non-contiguous `u` layout is covered.
- Black 26.3.1, SPDX, Python syntax, and whitespace checks pass.

## Summary by CodeRabbit

* **New Features**
  * Added an experimental HSTU LayerNorm-Multiply-SiLU-Dropout API with fused forward and explicit backward operations.
  * Added support for metadata-only tensor descriptors, dynamic row counts, stream-aware execution, and packed dropout masks on supported hardware.
  * Added HSTU LMSD benchmarks with predefined workload shapes, forward, backward, and end-to-end modes, configurable runs, JSON output, and shape listing.

* **Documentation**
  * Added API usage, configuration, installation, stream, caching, workspace, and benchmark guidance.

### CODE REVIEW

File: python/cudnn/hstu_lmsd/cutedsl/cute_dsl_ln_mul_dropout.py [hunk 1]
```diff
@@ -0,0 +1,674 @@
...
+    applied independently to each of the three segments. mask is (N, D) int8
+    with bit0=y, bit1=x, bit2=u; a set bit means keep. This layout is the
+    contract with the Triton backward -- see where the mask is read in
+    triton_hstu_linear.py.
+    """
+    if x.ndim != 2:
+        raise ValueError(f"x must be rank 2, got shape {tuple(x.shape)}")
+    N, D = x.shape
+    if not 1 <= N < (1 << 31):
+        raise ValueError(f"N must fit in a positive signed Int32, got {N}")
```
[the remark this answers is not in the document]
Comment (z52527) [reply python/cudnn/hstu_lmsd/cutedsl/cute_dsl_ln_mul_dropout.py]: Addressed in 26fe96fe. The original code path was removed or reorganized, and the corresponding validation/documentation is present in the current implementation.

File: python/cudnn/hstu_lmsd/cutedsl/cute_dsl_ln_mul_dropout_bwd.py [hunk 2]
```diff
@@ -0,0 +1,630 @@
...
+    if tuple(x.stride()) != (D, 1) or tuple(random_mask.stride()) != (D, 1):
+        raise ValueError(f"x and random_mask must have stride ({D}, 1)")
+    if dy.stride(1) != 1 or dy.stride(0) < 3 * D or dy.stride(0) % 8 != 0:
+        raise ValueError("dy must have unit innermost stride and 16-byte-aligned, non-overlapping rows")
+    if u.stride(1) != 1 or u.stride(0) < D or u.stride(0) % 8 != 0:
+        raise ValueError("u must have unit innermost stride and 16-byte-aligned, non-overlapping rows")
+    if tuple(weight.stride()) != (1,) or tuple(bias.stride()) != (1,) or tuple(mean.stride()) != (1,) or tuple(rstd.stride()) != (1,):
+        raise ValueError("weight, bias, mean, and rstd must be contiguous")
+    if any(tensor.data_ptr() % ALIGN != 0 for tensor in (dy, x, u, weight, bias, mean, rstd, random_mask)):
+        raise ValueError(f"all inputs must be {ALIGN}-byte aligned")
```
[the remark this answers is not in the document]
Comment (z52527) [reply python/cudnn/hstu_lmsd/cutedsl/cute_dsl_ln_mul_dropout_bwd.py]: Addressed in 26fe96fe. The original code path was removed or reorganized, and the corresponding validation/documentation is present in the current implementation.

File: test/python/hstu_lmsd/test_usage.py [hunk 3]
```diff
@@ -0,0 +1,75 @@
+# SPDX-FileCopyrightText: Copyright (c) 2026 NVIDIA CORPORATION & AFFILIATES. All rights reserved.
+# SPDX-License-Identifier: Apache-2.0
+
+"""Minimal public-API usage tests for HSTU LMSD."""
```
[the remark this answers is not in the document]
Comment (z52527) [reply test/python/hstu_lmsd/test_usage.py]: Addressed in 26fe96fe. The original code path was removed or reorganized, and the corresponding validation/documentation is present in the current implementation.

**Comment by jiayus-nvidia** [comment]
Covered by #946
